crates/core/src/factories/core_space.rs

```diff
@@ -92,13 +134,142 @@ impl Space for CoreSpace {
+            local_agent.set_tgt_storage_arc_hint(DhtArc::Empty);
+            local_agent.set_cur_storage_arc(DhtArc::Empty);
+
+            // update our local map
+            self.inner
+                .lock()
+                .unwrap()
+                .local_agent_map
+                .insert(local_agent.agent().clone(), local_agent.clone());
+
+            // TODO - inform gossip module of new join
+            // TODO - inform sharding module of new join
+
+            let inner = self.inner.clone();
+            let space = self.space.clone();
+            let local_agent2 = local_agent.clone();
+            let peer_store = self.peer_store.clone();
+            local_agent.register_cb(Arc::new(move || {
+                let inner = inner.clone();
+                let space = space.clone();
+                let local_agent2 = local_agent2.clone();
+                let peer_store = peer_store.clone();
+                tokio::task::spawn(async move {
+                    // TODO - call an update function on the gossip module.
+                    // TODO - call an update function on the sharding module.
+
+                    let url = inner.lock().unwrap().current_url.clone();
+
+                    if let Some(url) = url {
+                        // sign a new agent info
+                        let created_at = Timestamp::now();
+                        let expires_at = created_at
+                            + std::time::Duration::from_secs(60 * 20);
+                        let info = agent::AgentInfo {
+                            agent: local_agent2.agent().clone(),
+                            space,
+                            created_at,
+                            expires_at,
+                            is_tombstone: false,
+                            url: Some(url),
+                            storage_arc: local_agent2.get_cur_storage_arc(),
+                        };
+
+                        let info = match agent::AgentInfoSigned::sign(
+                            &local_agent2,
+                            info,
+                        )
+                        .await
+                        {
+                            Err(err) => {
+                                tracing::warn!(
+                                    ?err,
+                                    "failed to sign agent info",
+                                );
+                                return;
+                            }
+                            Ok(info) => info,
+                        };
+
+                        // add it to the peer_store.
+                        if let Err(err) = peer_store.insert(vec![info]).await {
+                            tracing::warn!(
+                                ?err,
+                                "failed to add agent info to peer store"
+                            );
+                        }
+
+                        // TODO - send the new agent info to bootstrap.
```

Comment:
Depends on the bootstrap client which has been merged right? Is this TODO just being left to keep this review under control or was it just not possible yet when the TODO was written?

Reply:
Correct, this means notify the bootstrap module, the module itself will decide if it sends to the server or not.

The PR was getting crazy, I just picked an arbitrary place to break off new work : )